proposals/opt-in.md

```diff
@@ -0,0 +1,289 @@
+
+1. If a declaration is annotated with the marker, it requires opt-in to that marker and can use other API *with that same marker* in its body.
+2. If a declaration or an expression is annotated with `@OptIn(Marker::class)`, it can use other declarations that use the selected marker, but it does not require opt-in itself (its clients will not *have* to opt in).
+
+The first option of usage of opt-in requirement markers is called a **propagating opt-in** (the annotation effectively causes propagation of the requirement), and the second — a **non-propagating opt-in**. The user is free to choose whichever option is preferable in each scenario.
+
+Example:
+
+```kotlin
+// Library code:
+
+@RequiresOptIn
+annotation class ShinyNewAPI
+
+// Class Foo requires opt-in with marker ShinyNewAPI
+@ShinyNewAPI
+class Foo { ... }
+
+// Function bar requires opt-in with marker ShinyNewAPI
+@ShinyNewAPI
+fun Foo.bar() = ...
+
+// Usage:
+
+// Function useFooAndBar uses API marked with ShinyNewAPI and thus
+// is required to opt in to that marker. Here, we choose a propagating opt-in,
+// because the API is used in the function signature (parameter foo's type).
+// (But this is not enforced by the compiler in any way, we could've chosen
+// a non-propagating opt-in as well.)
+@ShinyNewAPI
+fun useFooAndBar(foo: Foo) {
+    foo.bar()
+}
+
+// Function doSomething uses API marked with ShinyNewAPI and is also
+// required to opt in. Here, we choose a non-propagating opt-in, because the
+// API is used in the function body and it should not concern our clients
+@OptIn(ShinyNewAPI::class)
+fun doSomething() {
+    val foo = Foo()
+    foo.bar()
+}
+```
+
+Note that by opting into the API with the propagating opt-in, `useFooAndBar` effectively requires opt-in itself (with the same marker). In theory, we could distinguish initial introduction of the API and its propagating usages, but it would complicate the proposal a bit and there doesn't seem to be much value in doing that.
+
+Both opt-in mechanisms allow to use the API for the selected markers anywhere in the parse tree lexically under the annotated element.
+
+Using `OptIn` with annotations that are *not* opt-in requirement markers has no effect and yields a compilation warning. (Note that this must not be an error because user code should not break once an annotation is no longer an opt-in requirement marker.) Using `OptIn` with no arguments has no effect and yields a warning as well.
+
+## Opt-in requirement message
+
+When using an API that requires opt-in without the said opt-in, the compiler reports a warning or an error depending on the specified level. It's possible to specify a custom message in `RequiresOptIn`, that will be reported by the compiler. If no message is given (i.e. if `message` is empty), the compiler will report that the API is *experimental*:
+
+```
+test.kt:22:9: error: this declaration is experimental and its usage must be marked with '@ShinyNewAPI' or '@OptIn(ShinyNewAPI::class)'
```

Comment:
There's a separate discussion in https://youtrack.jetbrains.com/issue/KT-47925 on making a better message and it's great to correct it in this KEEP, too.